### PR TITLE
fix nested column thread safety issue for 24.0.1

### PR DESCRIPTION
**Pull request by clintropolis**
smaller footprint version of fix #13265 for 24.0.1, this just fixes the `TypeStrategy` to use positional reads

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java [hunk 1]
```diff
@@ -163,13 +162,8 @@ private GlobalDictionarySortedCollector getSortedIndexFromIncrementalAdapter(
...
     final NestedDataColumnIndexer indexer = (NestedDataColumnIndexer) dim.getIndexer();
-    for (Map.Entry<String, NestedDataColumnIndexer.LiteralFieldIndexer> entry : indexer.fieldIndexers.entrySet()) {
-      // skip adding the field if no types are in the set, meaning only null values have been processed
-      if (!entry.getValue().getTypes().isEmpty()) {
-        mergedFields.put(entry.getKey(), entry.getValue().getTypes());
-      }
-    }
-    return indexer.globalDictionary.getSortedCollector();
+    indexer.mergeFields(mergedFields);
+    return indexer.getSortedCollector();
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:166]: This change is just general refactoring and not actually functionally meaningful, right?  Just validating my understanding

**Review by imply-cheddar — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:166]: yep, made writing the test easier and the direct field access was kind of ugly

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by xvrl** [comment]
@clintropolis it looks like integration tests are still failing in the 24.0.1 branch. We might want to check if there is something non-deterministic or not.

**Comment by clintropolis** [comment]
> it looks like integration tests are still failing in the 24.0.1 branch. We might want to check if there is something non-deterministic or not.

hmm, these actually look unrelated to my fix, but maybe worth looking into

```
[ERROR] org.apache.druid.sql.avatica.DruidAvaticaHandlerTest.testConcurrentQueries  Time elapsed: 12.981 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.apache.calcite.avatica.AvaticaSqlException: Error -1 (00000) : Error while executing SQL "SELECT COUNT(*) + 1636 AS ci FROM foo": Remote driver error: QueryInterruptedException: Error while applying rule DruidQueryRule(AGGREGATE_PROJECT), args [rel#631993:LogicalProject.NONE.[](input=RelSubset#631992,ci=+($0, 1636)), rel#632012:DruidQueryRel.NONE.[](query={"queryType":"timeseries","dataSource":{"type":"table","name":"foo"},"intervals":{"type":"intervals","intervals":["-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"]},"granularity":{"type":"all"},"aggregations":[{"type":"count","name":"a0"}],"context":{"nativeQueryIds":"[3ef1e7e1-f2f6-4152-8384-229eaea39482]","sqlQueryId":"f2727102-ff6c-404c-8f9a-2ed84a406e46","sqlStringifyArrays":false}},signature={a0:LONG})] -> RuntimeException: Error while applying rule DruidQueryRule(AGGREGATE_PROJECT), args [rel#631993:LogicalProject.NONE.[](input=RelSubset#631992,ci=+($0, 1636)), rel#632012:DruidQueryRel.NONE.[](query={"queryType":"timeseries","dataSource":{"type":"table","name":"foo"},"intervals":{"type":"intervals","intervals":["-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"]},"granularity":{"type":"all"},"aggregations":[{"type":"count","name":"a0"}],"context":{"nativeQueryIds":"[3ef1e7e1-f2f6-4152-8384-229eaea39482]","sqlQueryId":"f2727102-ff6c-404c-8f9a-2ed84a406e46","sqlStringifyArrays":false}},signature={a0:LONG})] -> NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
```

I think maybe https://github.com/apache/druid/pull/13049 is the fix

**Comment by clintropolis** [comment]
created #13281 to backport that fix

**Pull request merged**